File: courses/focs-fa20/homeworks/homework3.ml
(* 

HOMEWORK 3

Due: Thu Oct 15, 2020 (23h59)

Name: 

Email:

Remarks, if any:

*)


(*
 *
 * Please fill in this file with your solutions and submit it
 *
 * The functions below are stubs that you should replace with your
 * own implementation.
 *
 * PLEASE DO NOT CHANGE THE TYPES IN THE STUBS BELOW.
 * Doing so risks making it impossible for me to test your code.
 *
 * Always make sure you can #use this file in a FRESH OCaml shell
 * before submitting it. It has to load without any errors.
 *
 *)


(* 
 * String <-> characters utility functions:
 *
 *   explode : string -> char list
 *      returns the list of characters making up a string
 *
 *   implode : char list -> string
 *      concatenates the list of characters into a string
 *
 *)

let explode (str) = 
  let rec acc (index,result) = 
    if (index<0) then
      result
    else
      acc(index-1, (String.get str index)::result)
  in
    acc(String.length(str)-1, [])

let implode cs = 
  List.fold_right (fun a r -> (String.make 1 a)^r) cs ""


(*
 *  The type of a deterministic finite automaton
 * 
 *)

type dfa = { states: int list;
     	     alphabet: char list;
             delta: int -> char -> int;
             start : int;
             final : int list }


(* 
 * Sample Deterministic Finite Automata
 *
 * The first accepts the language of all strings over {x,y} 
 * with a multiple-of-3 number of x's.
 *
 * The second accepts the language of all strings over {x,y} 
 * starting and ending with an x.
 *
 *)

let dfa1 = { 
  states = [1; 2; 3];
  alphabet = ['x'; 'y'];
  delta = (fun q x -> match (q, x) with
                      | (1, 'x') -> 2
	              | (2, 'x') -> 3
	              | (3, 'x') -> 1
	              | (1, 'y') -> 1
	              | (2, 'y') -> 2
	              | (3, 'y') -> 3);
  start = 1;
  final = [1]
} 


let dfa2 = { 
  states = [10; 11; 12; 13];
  alphabet = ['x'; 'y'];
  delta = (fun q x -> match (q, x) with
                      | (10, 'x') -> 11
	              | (11, 'x') -> 12
	              | (11, 'y') -> 11
	              | (12, 'x') -> 12
	              | (12, 'y') -> 11
	              | (_, _) -> 13);
  start = 10;
  final = [12]
} 


(* QUESTION 1 *)

let absNonZero (xs: int list): int list =
  failwith "Not implemented"

  
let map_functions (fs: ('a -> 'b) list) (x: 'a): 'b list =
  failwith "Not implemented"

  
let compose_all (fs: ('a -> 'a) list) (x: 'a): 'a =
  failwith "Not implemented"


let pairs1 (x: 'a) (ys: 'b list): ('a * 'b) list =
  failwith "Not implemented"

  
let pairs (xs: 'a list) (ys: 'b list): ('a * 'b) list =
  failwith "Not implemented"



(* Question 2 *)

let cons_all (x: 'a) (xss: 'a list list): 'a list list = 
  failwith "Not implemented"
    

let prefixes (xs: 'a list): 'a list list = 
  failwith "Not implemented"

  
let suffixes (xs: 'a list): 'a list list =
  failwith "Not implemented"


let splits (xs: 'a list): ('a list * 'a list) list =
  failwith "Not implemented"

  
let inject (x: 'a) (xs: 'a list): 'a list list =
  failwith "Not implemented"

  
(* BONUS *)

let permutations (xs: 'a list): 'a list list = 
  failwith "Not implemented"



   
(* Question 3 *)

let accept (m: dfa) (input: string): bool =
  failwith "Not implemented"



  

  
(* This function is the base function that basically loops through all
 * strings  of length up to n, and prints those that are accepted by the
 * finite automaton.
 *
 * This is being way too clever to try to not blow the stack 
 * while enumerating all strings up to a given length. Basically.
 * we enumerate all integer, convert them to base K (where K is the
 * size of the alphabet) and then replace every digit base K by the
 * letter of the alphabet at the corresponding index in the alphabet. 
 *
 * The key is that we can enumerate integers super easily
 *
 *)

let lang (m: dfa) (n: int) = 

  let rec expt a n = if n <= 0 then 1 else a*(expt a (n-1)) in
  
  let rec take n default l = 
    if n <= 0 then []
    else (match l with
          | [] -> default::(take (n-1) default l)
          | x::xs -> x::(take (n-1) default xs)) in
  
  let to_base_n base size n = 
    let rec loop n = 
      if n <= 0 then []
      else if n mod base = 0 then 0::(loop (n / base))
      else (n mod base)::(loop ((n - n mod base) / base))  in
    take size 0 (loop n)  in
  
  let to_string alphabet size n = 
    let base = List.length alphabet in
    let num_base = to_base_n base size n in
    implode (List.map (fun i -> List.nth alphabet i) num_base) in
  
    if n < 0 then ()
    else
      let print_str s = if s = "" then print_string "  <epsilon>\n"
  	              else print_string ("  "^s^"\n")  in
      let rec loop i = 
        if i <= n then 
  	  let ts = to_string m.alphabet i  in
  	  let bound = expt (List.length m.alphabet) i in
  	  let rec loop2 j = 
  	    if j < bound then (if accept m (ts j) 
                                 then print_str (ts j)
                               else ();
  			       loop2 (j+1))
  	    else ()  in
  	  (loop2 0; loop (i+1))
        else ()  in
      loop 0

File: courses/focs-fa24/homeworks/3/homework3.py
# --------------------------------------------------
#
# HOMEWORK 3
#
# Due: Sun, Oct 13 2024 (23h59)
#
# Name:
#
# Email:
#
# Remarks, if any:
#
#
#
# --------------------------------------------------
#
# Please fill in this file with your solutions and submit it
#
# The functions below are stubs that you should replace with your
# own implementation.
#
# PLEASE DO NOT CHANGE THE SIGNATURE IN THE STUBS BELOW.
# Doing so risks making it impossible for me to test your code.
#
# --------------------------------------------------



#
# Some sample PDAs.
#

PDA_ANBN = {
    "states": [0, 1],
    "alphabet": ["a", "b"],
    "stack_alphabet": [".", "X"],
    "start": 0,
    "final": [1],
    "delta": [
        (0, "a", ".", 0, ["X", "."]),
        (0, "a", "X", 0, ["X", "X"]),
        (0, None, ".", 1, ["."]),
        (0, None, "X", 1, ["X"]),
        (1, "b", "X", 1, []),
    ]
}

PDA_PAL = {
    "states": [0, 1],
    "alphabet": ["a", "b"],
    "stack_alphabet": [".", "X", "Y"],
    "start": 0,
    "final": [1],
    "delta": [
        (0, "a", ".", 0, ["X", "."]),
        (0, "a", "X", 0, ["X", "X"]),
        (0, "a", "Y", 0, ["X", "Y"]),
        (0, "b", ".", 0, ["Y", "."]),
        (0, "b", "X", 0, ["Y", "X"]),
        (0, "b", "Y", 0, ["Y", "Y"]),
        (0, None, ".", 1, ["."]),
        (0, None, "X", 1, ["X"]),
        (0, None, "Y", 1, ["Y"]),
        (0, "a", ".", 1, ["."]),
        (0, "a", "X", 1, ["X"]),
        (0, "a", "Y", 1, ["Y"]),
        (0, "b", ".", 1, ["."]),
        (0, "b", "X", 1, ["X"]),
        (0, "b", "Y", 1, ["Y"]),
        (1, "a", "X", 1, []),
        (1, "b", "Y", 1, [])
    ]
}

# A _deterministic_ PDA corresponding to PDA_ANBN.

DPDA_ANBN = {
    "states": [0, 1, 2, 99],
    "alphabet": ["a", "b"],
    "stack_alphabet": [".", "X"],
    "start": 0,
    "final": [0, 2],
    "delta": [
        (0, "a", ".", 1, ["X", "."]),
        (1, "a", "X", 1, ["X", "X"]),
        (1, "b", "X", 2, []),
        (2, "b", "X", 2, []),

        # Sink transitions (for deterministism)
        (0, "a", "X", 99, []),
        (0, "b", ".", 99, ["."]),
        (0, "b", "X", 99, []),
        (1, "a", ".", 99, ["."]),
        (1, "b", ".", 99, ["."]),
        (2, "a", ".", 99, ["."]),
        (2, "a", "X", 99, []),
        (2, "b", ".", 99, ["."]),
        (99, "a", ".", 99, ["."]),
        (99, "a", "X", 99, []),
        (99, "b", ".", 99, ["."]),
        (99, "b", "X", 99, [])
    ]
}


#
# A generic function to create an acceptance function for nondeterministic machines.
#

def accept_ndet(step, start, is_done, is_accept):
    def accept(m, input):
        config = start(m, input)
        seen = []
        queue = [config]
        while queue:
            config = queue[0]
            queue = queue[1:]
            if config in seen:
                continue
            seen.append(config)
            if is_done(m, config):
                if is_accept(m, config):
                    return True
            next = step(m, config)
            queue.extend(next)
        return False
    return accept


#
# Function to print the language of an automaton
# up to a certain string length.
#

def language(m, n, accept):

    def to_base_n_rev(num, n):
        result = []
        while num > 0:
            result.append(num % n)
            num = num // n
        return result

    def enum(size):
        result = []
        for num in range(len(alphabet) ** size):
            digits = to_base_n_rev(num, len(alphabet))
            digits.extend([0] * (size - len(digits)))
            result.append("".join(alphabet[digits[size - i - 1]] for i in range(size)))
        return result

    alphabet = m["alphabet"]
    for size in range(n + 1):
        for input in enum(size):
            if accept(m, input):
                print(input or "<empty string>")


#############################################################
#
# Question 1
#

def step_pda(m, config):
    raise Exception("Not implemented")


def start_pda(m, input):
    raise Exception("Not implemented")


def is_done_pda(m, config):
    raise Exception("Not implemented")


def is_accept_pda(m, config):
    raise Exception("Not implemented")


def accept_dpda(m, input):
    raise Exception("Not implemented")



#############################################################
#
# Question 2
#

#
# The acceptance function for nondeterministic PDAs.
#

accept_pda = accept_ndet(step_pda, start_pda, is_done_pda, is_accept_pda)


PDA_AMBN_MGN = None


PDA_AMBN_NGM = None


PDA_AMBNCMN = None


PDA_AMBMNCN = None
